Create airgapped-wallet.py (#1558)
XRPLF/xrpl-dev-portal: content/_code-samples/airgapped-wallet/py/README.md

*Note*: You should not use this airgapped wallet in production, it should only be used for educational purposes only. 

This code sample consists of 2 parts:

- `airgapped-wallet.py` - This code should be stored in a standalone airgapped machine, it consist of features to generate a wallet, store a keypair securely, sign a transaction and share the signed transaction via QR code.
- `relay-transaction.py` - This code could be stored in any online machine, no credentials is stored on this code other than a signed transaction which would be sent to an XRPL node for it to be validated on the ledger.

Preferably, `airgapped-wallet.py` should be on a Linux machine while `relay-transaction.py` could be on any operating system.

# Security Practices
Strongly note that an airgapped system's security is not determined by its code alone but the security practices that are being followed by an operator.

There are channels that can be maliciously used by outside parties to infiltrate an airgapped system and steal sensitive information.

> There are other ways malware could interact across airgapped networks, but they all involve an infected USB drive or a similar device introducing malware onto the airgapped machine. (They could also involve a person physically accessing the computer, compromising it and installing malware or modifying its hardware.)

Reviewer: nit: No need for parens here since you end with a period anyway. 
```suggestion
There are other ways malware could interact across airgapped networks, but they all involve an infected USB drive or a similar device introducing malware onto the airgapped machine. They could also involve a person physically accessing the computer, compromising it and installing malware or modifying its hardware.
```

# Tutorial
For testing purposes, you would need to have 2 machines and 1 phone in hand to scan the QR code.

1. 1st machine would be airgapped, following the security practices written [here](#security-practices). It stores and manages an XRPL Wallet. 
2. 2nd machine would be a normal computer connected to the internet. It relays a signed transaction blob to a rippled node.
3. The phone would be used to scan a QR code, which contains a signed transaction blob. The phone would transmit it to the 2nd machine.

The diagram below shows you the process of submitting a transaction to the XRPL:
<p align="center">
  <img src="https://user-images.githubusercontent.com/87929946/197970678-2a1b7f7e-d91e-424e-915e-5ba7d34689cc.png" width=75% height=75%>
</p>

# Setup
- Machine 1 - An airgapped computer (during setup, it must be connected to the internet to download the files)
- Machine 2 - A normal computer connected to the internet
- Phone - A normal phone with a working camera to scan a QR

## Machine 1 Setup

> Since this machine will be airgapped, it is best to use Linux as the Operating System.

Reviewer: I'm not clear on why that is - is it because Linux has better stability without updates? Better in-built security?

Author: As compared to Windows, Linux is less prone to malware since a huge number of viruses are developed specifically for Windows, most viruses can't harm Linux in this case.

1. Install Python 3.8:

**Linux Command Line**:
```
sudo apt-get update
sudo apt-get install python3.8 python3-pip
```
**Website**: https://www.python.org/downloads/source/

> 2. Clone all the files under the [`airgapped-wallet`](https://github.com/XRPLF/xrpl-dev-portal/tree/master/content/_code-samples/airgapped-wallet/py) directory

Reviewer: Is there an easily copy + pastable command you could add here for that?

Author: I don't think so, I tried using git to only clone the files under the `airgapped-wallet` directory but git could only clone the entire repo (IIRC).

Reviewer: Maybe better phrasing would be `Download` instead of `Clone` then? (Because otherwise I feel obligated to google it and then I'm in a rabbit hole where Git doesn't do what I want)

Author: Right.

1. Install Python 3.8

**Linux Command Line**:
```
sudo apt-get update
sudo apt-get install python3.8 python3-pip
```
**Website**: https://www.python.org/downloads/source/

2. Clone all the files under the [`airgapped-wallet`](https://github.com/XRPLF/xrpl-dev-portal/tree/master/content/_code-samples/airgapped-wallet/py) directory

3. Import all the modules required by running:
```
pip install -r requirements.txt
```

4. Edit line 47 @ `relay-transaction.py` and insert the signed transaction blob

5. Run `relay-transaction.py`

## Phone Setup
The phone requires a working camera that is able to scan a QR code and an internet connection for it to be able to transmit the signed transaction blob to Machine 2.

Once you have signed a transaction in the airgapped machine, a QR code will be generated which will contain the signed transaction blob. Example:

<img src="https://user-images.githubusercontent.com/87929946/196018292-f210a9f2-c5f8-412e-98c1-361a72286378.png" width=20% height=20%>

Scan the QR code using the phone and transmit it to Machine 2, which will then be sending it to a rippled node.

> You could use Discord, Whatsapp or even e-mail the signed transaction blob to yourself in order for you to receive the signed transaction blob in Machine 2.

Reviewer: Slight rephrasing suggestion
```suggestion
You can send a message to yourself using Discord, WhatsApp or even e-mail, then open up the message using Machine 2 to receive the signed transaction blob.
```

The diagram below shows you the process of submitting a transaction to the XRPL:
<p align="center">
  <img src="https://user-images.githubusercontent.com/87929946/197970678-2a1b7f7e-d91e-424e-915e-5ba7d34689cc.png" width=75% height=75%>
</p>

# Setup
- Machine 1 - An airgapped computer (during setup, it must be connected to the internet to download the files)
- Machine 2 - A normal computer connected to the internet
- Phone - A normal phone with a working camera to scan a QR

## Machine 1 Setup
Since this machine will be airgapped, it is best to use Linux as the Operating System.

1. Install Python 3.8:

**Linux Command Line**:
```
sudo apt-get update
sudo apt-get install python3.8 python3-pip
```
**Website**: https://www.python.org/downloads/source/

2. Clone all the files under the [`airgapped-wallet`](https://github.com/XRPLF/xrpl-dev-portal/tree/master/content/_code-samples/airgapped-wallet/py) directory

3. Import all the modules required by running:
```
pip install -r requirements.txt
```

4. Airgap the machine by following the security practices written [here](#security-practices).

> 5. Run `airgapped-machine.py`

Reviewer: There is no `airgapped-machine.py` - should this be `airgapped-wallet.py`?

Author: Yup, my mistake.

> 6. Scan the QR code and fund the account using the [testnet faucet](https://test.bithomp.com/faucet/) 

Reviewer: ```suggestion
6. Scan the QR code and fund the account using the [testnet faucet](https://test.bithomp.com/faucet/) 

7. Re-run the script and input '1' to generate a new transaction by following the instructions.

8. Use your phone to scan the QR code, then to send the signed transaction to Machine 2 for submission
```

## Machine 2 Setup
This machine will be used to transmit a signed transaction blob from Machine 1, it would require internet access.

1. Install Python 3.8

**Linux Command Line**:
```
sudo apt-get update
sudo apt-get install python3.8 python3-pip
```
**Website**: https://www.python.org/downloads/source/

2. Clone all the files under the [`airgapped-wallet`](https://github.com/XRPLF/xrpl-dev-portal/tree/master/content/_code-samples/airgapped-wallet/py) directory

3. Import all the modules required by running:
```
pip install -r requirements.txt
```

> 4. Edit line 47 @ `relay-transaction.py` and insert the signed transaction blob

Reviewer: Calling back so people know where to look if they missed a step.

```suggestion
4. Edit line 47 @ `relay-transaction.py` and insert the signed transaction blob from scanning the QR code Machine 1 generated.
```

> 4. Edit line 47 @ `relay-transaction.py` and insert the signed transaction blob

Reviewer: I'm trying to follow the steps, but I'm confused where we get a signed transaction blob - seems like a couple steps are missing, but I'm not sure what they are ("Maybe grab the wallet generated via the QR code and use it in Xumm to sign an accountset transaction or something"?)

Reviewer: Or is there supposed to be a QR code I can scan with the transaction on it? When running the steps I just get one that gives me an address as far as I can tell

Author: There's no use of XUMM in this code sample, after you've signed a transaction blob on the airgapped machine:
A QR code will pop up and you'd need to scan it, the content is the signed transaction blob.

>When running the steps I just get one that gives me an address as far as I can tell

When you first run `airgapped-wallet.py`, it just sets up the files & wallet credentials. *This ^* QR code contains the wallet's address. You need to re-run `airgapped-wallet.py` to be able to sign transactions and such.

5. Run `relay-transaction.py`

> ## Phone Setup

Reviewer: I'd recommend removing this section, and just having the instructions be part of Machine 1 setup (Ex. of how that could be done in the above recommendation)

Reviewer: That way the tutorial has a linear flow from top to bottom with steps to do.

Author: I have to disagree here, I'm sure that there will definitely be users who aren't familiar with this.
